File: cpp/benchmarks/bench_heap.cpp
#include "bench_common.hpp"
#include "heap.hpp"
#include <chrono>
#include <random>

int main() {
    std::string out_dir = get_results_dir();
    fs::create_directories(out_dir);

    std::random_device rd;
    std::mt19937 g(rd());

    std::string csv_path = out_dir + "/cpp_heap.csv";
    std::ofstream file(csv_path);
    if (!file.is_open()) {
        std::cerr << "Failed to open " << csv_path << "\n";
        return 1;
    }
    file << "N,insert_mean_ms,insert_std_ms,get_mean_ms,get_std_ms,memory_mb\n";
    file << std::fixed << std::setprecision(6);

    for (int n : SCALES) {
        std::vector<int> keys(n);
        std::iota(keys.begin(), keys.end(), 0);
        std::shuffle(keys.begin(), keys.end(), g);

        {  // warm-up: insert = push, get = peek
            heap::Heap h;
            for (int k : keys) h.push(k);
            while (h.size() > 0) do_not_optimize(h.peek()), h.pop();
        }

        std::vector<double> insert_ms(NUM_RUNS), get_ms(NUM_RUNS);
        for (int run = 0; run < NUM_RUNS; run++) {
            std::shuffle(keys.begin(), keys.end(), g);
            heap::Heap h;
            auto start = std::chrono::high_resolution_clock::now();
            for (int k : keys) h.push(k);
            insert_ms[run] = std::chrono::duration<double, std::milli>(std::chrono::high_resolution_clock::now() - start).count();
            start = std::chrono::high_resolution_clock::now();
            for (size_t i = 0; i < static_cast<size_t>(n); i++) do_not_optimize(h.peek());
            get_ms[run] = std::chrono::duration<double, std::milli>(std::chrono::high_resolution_clock::now() - start).count();
        }
        double i_mean, i_std, g_mean, g_std;
        mean_std(insert_ms, i_mean, i_std);
        mean_std(get_ms, g_mean, g_std);
        double mem = memory_mb();
        file << n << "," << i_mean << "," << i_std << "," << g_mean << "," << g_std << "," << std::setprecision(4) << mem << "\n";
        file << std::setprecision(6);
        std::cout << "N=" << n << ": Insert " << i_mean << " ± " << i_std << " ms, Get(peek) " << g_mean << " ± " << g_std << " ms, memory=" << mem << " MB\n";
    }
    std::cout << "Wrote " << csv_path << "\n";
    return 0;
}

File: python/benchmarks/bench_heap.py
#!/usr/bin/env python3
"""Heap benchmark: push (insert) and pop (remove). Same schema as hashmap."""

import sys
import os
import random
import time

_script_dir = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, _script_dir)
sys.path.insert(0, os.path.join(_script_dir, ".."))
from src.heap import MinHeap
from bench_common import (
    SCALES,
    NUM_RUNS,
    get_memory_mb,
    mean_std,
    write_csv,
    get_results_dir,
)


def main():
    results_dir = get_results_dir()
    os.makedirs(results_dir, exist_ok=True)

    rows = [
        [
            "N",
            "insert_mean_ms",
            "insert_std_ms",
            "get_mean_ms",
            "get_std_ms",
            "memory_mb",
        ]
    ]
    for N in SCALES:
        keys = list(range(N))
        random.shuffle(keys)

        # warm-up
        h = MinHeap()
        for k in keys:
            h.insert(k)
        while h.size() > 0:
            _ = h.pop()

        insert_times = []
        pop_times = []
        for _ in range(NUM_RUNS):
            random.shuffle(keys)
            h = MinHeap()
            start = time.perf_counter()
            for k in keys:
                h.insert(k)
            insert_times.append((time.perf_counter() - start) * 1000)
            start = time.perf_counter()
            while h.size() > 0:
                _ = h.pop()
            pop_times.append((time.perf_counter() - start) * 1000)

        i_mean, i_std = mean_std(insert_times)
        p_mean, p_std = mean_std(pop_times)
        memory_mb = get_memory_mb()
        rows.append(
            [
                N,
                f"{i_mean:.6f}",
                f"{i_std:.6f}",
                f"{p_mean:.6f}",
                f"{p_std:.6f}",
                f"{memory_mb:.4f}",
            ]
        )
        print(
            f"N={N}: Insert {i_mean:.6f} ± {i_std:.6f} ms, Pop {p_mean:.6f} ± {p_std:.6f} ms, memory={memory_mb:.4f} MB"
        )
    write_csv(rows, results_dir, "python_heap.csv")


if __name__ == "__main__":
    main()
